File: intel_briefing/sensors/wallstreetcn.py
# ABOUTME: WallStreetCN (华尔街见闻) sensor using the public information-flow API.
# ABOUTME: Fetches global finance and macro news from China's leading financial news platform.
import logging
from datetime import datetime

import httpx

from intel_briefing.models import IntelItem, ConfigSettings

logger = logging.getLogger(__name__)

_WSCN_URL = (
    "https://api-one.wallstcn.com/apiv1/content/information-flow"
    "?channel=global-channel&accept=article&limit=30"
)


class WallStreetCNSensor:
    """Fetches latest finance and macro news from WallStreetCN."""

    sensor_name = "wallstreetcn"

    def fetch(self, config: ConfigSettings, limit: int) -> list[IntelItem]:
        try:
            resp = httpx.get(_WSCN_URL, timeout=15, follow_redirects=True)
            resp.raise_for_status()
            data = resp.json()
        except Exception as exc:
            logger.warning("WallStreetCN fetch failed: %s", exc)
            return []

        items: list[IntelItem] = []
        for raw in data.get("data", {}).get("items", []):
            if len(items) >= limit:
                break

            res = raw.get("resource")
            if not res:
                continue

            title = res.get("title") or res.get("content_short")
            if not title:
                continue

            url = res.get("uri") or ""
            ts = res.get("display_time", 0)
            time_str = datetime.fromtimestamp(ts).strftime("%Y-%m-%d %H:%M") if ts else None
            item_id = str(res.get("id") or hash(title) & 0xFFFF)

            items.append(
                IntelItem(
                    id=f"wscn-{item_id}",
                    source="wallstreetcn",
                    title=title,
                    url=url,
                    published_at=time_str,
                )
            )

        logger.info("WallStreetCN sensor returned %d items", len(items))
        return items

File: frontend/src/lib/sensors/wallstreetcn.ts
// ABOUTME: WallStreetCN sensor using the public information-flow API.
// ABOUTME: Fetches global finance and macro news from China's leading financial news platform.
import type { ConfigSettings, IntelItem } from '../models'

const WSCN_URL =
  'https://api-one.wallstcn.com/apiv1/content/information-flow?channel=global-channel&accept=article&limit=30'

export async function fetchWallStreetCN(_config: ConfigSettings, limit: number): Promise<IntelItem[]> {
  try {
    const resp = await fetch(WSCN_URL, {
      signal: AbortSignal.timeout(15000),
      redirect: 'follow',
    })
    if (!resp.ok) throw new Error(`HTTP ${resp.status} from WallStreetCN`)
    const data = await resp.json() as Record<string, unknown>

    const items: IntelItem[] = []
    const rawItems = ((data.data as Record<string, unknown>)?.items as Array<Record<string, unknown>>) ?? []

    for (const raw of rawItems) {
      if (items.length >= limit) break
      const res = raw.resource as Record<string, unknown> | undefined
      if (!res) continue

      const title = String(res.title ?? res.content_short ?? '')
      if (!title) continue

      const url = String(res.uri ?? '')
      const ts = Number(res.display_time ?? 0)
      const timeStr = ts ? new Date(ts * 1000).toISOString().slice(0, 16).replace('T', ' ') : null
      const itemId = String(res.id ?? (Math.abs(hashString(title)) & 0xFFFF))

      items.push({
        id: `wscn-${itemId}`,
        source: 'wallstreetcn',
        title,
        url,
        published_at: timeStr,
      })
    }
    return items
  } catch (err) {
    throw err instanceof Error ? err : new Error(String(err))
  }
}

function hashString(s: string): number {
  let hash = 0
  for (let i = 0; i < s.length; i++) {
    hash = ((hash << 5) - hash + s.charCodeAt(i)) | 0
  }
  return hash
}
